### DISCUSSION TITLE
iOS Static vs Dynamic Linking

### DISCUSSION THREAD

**Discussion by omarzl** [MobileNativeFoundation/discussions#133] [General]
This is a topic that has many points of view, I will like to discuss the pros and cons other companies have determined and which linking are using.

In our case we migrated from dynamic to static linking around 2 years ago, the main decision was the startup time, we were in around 8 secs and today we are in around 1.5 secs, others have seen this notable reduction too?

In that time we weren't tracking the app size so I don't know if it benefit us but I have noticed that when we integrate 3rd party sdks and compare their static and dynamic versions, the static one has a smaller size impact, AFAIK this is because during the linking phase only the used symbols are linked into the binary while in the dynamic way it is copied the whole library/framework into the .app, has anyone noticed the opposite behaviour?

One downside of using static linking is, well the linking phase, since it needs to "join" all the symbols into one it becomes one of our bottlenecks in incremental compilation, we have consider having dynamic linking for local compilation and static linking for release builds but it is not as trivial configuration for us as we would like.

It is important to mention that we have around 180 static and 10 dynamic libraries at this moment.

Hope to hear other's opinions!

**Comment by tinder-maxwellelliott** [comment]
> In our case we migrated from dynamic to static linking around 2 years ago, the main decision was the startup time, we were in around 8 secs and today we are in around 1.5 secs, others have seen this notable reduction too?

We have seen around a 10x reduction in startup time from static linking, from 3 seconds to .3 seconds

> In that time we weren't tracking the app size so I don't know if it benefit us but I have noticed that when we integrate 3rd party sdks and compare their static and dynamic versions, the static one has a smaller size impact, AFAIK this is because during the linking phase only the used symbols are linked into the binary while in the dynamic way it is copied the whole library/framework into the .app, has anyone noticed the opposite behavior?

Seeing the same behavior, seeing around a 9MB reduction in the final shipped IPA

> One downside of using static linking is, well the linking phase, since it needs to "join" all the symbols into one it becomes one of our bottlenecks in incremental compilation, we have consider having dynamic linking for local compilation and static linking for release builds but it is not as trivial configuration for us as we would like.

We are looking into dynamically flipping out our linking strategy locally vs in CI to accomplish this. Bazel + Code gened BUILD files makes this easy for us

**Reply by mattrobmattrob** [reply]
How are you all doing (or have you investigated) for allowing dynamic framework/bundle/etc. access for resources (strings, images, asset catalogs, etc.) in both the static and dynamically linked scenarios?

Creating something similar to this, has helped with migration from dynamic to static for me with multiple companies in the past:
```Swift
extension Bundle {
  private final class FooBundleProvider {}

  static let fooBundle: Bundle = {
    // `FooBundleProvider` will be in the framework when it's dynamic or
    // in the app/test bundle when it's static. The placement of the bundle
    // should match.
    let bundle = Bundle(for: FooBundleProvider)
    let path = bundle.url(forResource: "FooBundle", withExtension: "bundle")!
    return Bundle(url: path)!
  }()
}
```
But like the comment says, it has the expectation that the bundle is co-located next to the binary that the source is linked into.

**Reply by keith** [reply]
At Lyft we pair each module with a generated source file that has some code like you posted, so for our approach to swapping out static libs for dylibs we would just update that code as needed. And just all devs only ever access resources through helpers that go through that code.

**Comment by keith** [comment]
Here's my attempt to summarize the pros for each option (the cons are kinda just the opposite of the pros for the other option):

Dynamic linking pros:

- More parallel, so likely faster, link times for modular projects. Because each library can be linked individually they can be done in parallel, and don't need to be re-linked unless you change some code that actually causes them to be rebuilt
- Smaller links, since each library only contains a subset of your entire project, each individual link likely takes significantly less time than an entirely static link of a large project
- Shared between extension targets. If you have app extensions, you can share the same dynamic binaries between your primary app target and the extensions, saving app download size, vs duplicating them by linking them for each final binary
- Deferred launch time loading. If you want to get pretty advanced you can defer loading until the library is actually needed with manually `dlopen`ing them, which can provide some safety from `+load` crashes, and less impact overall on launch time if the libraries aren't in the critical path of most sessions

Static linking pros:

- Can be dead stripped at link time. Since dynamic library linking happens without the full context of the final binary, the linker cannot know what symbols are eventually unused, where as with static linking it can, and therefore it can save you some binary size. This is especially important if you build many apps which could only use different subsets of the code from the same libraries.
- Lower launch time impact. Since dyld doesn't have to resolve your statically linked symbols at launch time, this saves some work there. Note that in advanced setups if your binary is quite large, at least in the past, you could hit other issues here since the entire binary has to be loaded into memory to validate code signatures at launch time, this lead some folks to use the `dlopen` path described above where possible

Some other notes:

- There is a hard limit to the number of dynamic libraries your binary can load, even ignoring the launch time hit. The current number is up to 512 before it hard crashes on launch, this is probably enough for a while, but not for us anymore 😬
- There is a significant impact on lldb attach time when you have many dylibs, I noticed about a 4x increase in lldb attach time on our project in light testing
- I think a lot of folks are trying the "dynamic for local, static for release" approach above, and I think that's a great path, especially for test bundles when you have many of those, 1 per each library perhaps, we have huge issues with disk space and this today when you have many test bundles linking all the same helper libraries
- Some folks have tried a bit more of a "mono dylib" approach to avoid some of the issues with launch time, but being smart about which libraries to include in that to balance when you have to relink based on what the developer is changing can be hard
- There are quite a few things you can do to improve link times in general, you can use [zld](https://github.com/michaeleisel/zld/) as a replacement linker (we use it but only for debug builds today), I'm also optimistic about [mold](https://github.com/rui314/mold) where macho support is currently being worked on. Even if you don't want to use a replacement linker there are some good performance tips captured [here](https://github.com/michaeleisel/zld/#other-things-to-speed-up-linking)
- We are on the order of thousands of static libs around (not all are linked into every app), and just 2 or so dynamic libs from 3rd party vendors

**Reply by omarzl** [reply]
Regarding the "mono dylib" I remembered this repo https://github.com/grab/cocoapods-pod-merge, its actually specific for CocoaPods but the idea is the same, merging modules into one dylib to reduce its launch time impact

**Reply by ferologics** [reply]
Great summary @keith! Thanks a lot 🙏

**Reply by annomusa** [reply]
"mono dylib" can be implemented by using SPM too anyway, it's better since it doesn't need to change import signature.
I think it's using export import of the dependenicies
https://github.com/renaudjenny/Swift-Package-Manager-Static-Dynamic-Xcode-Bug

**Comment by qyang-nj** [comment]
@keith summarized this really well 🙏. I just want to add a few points.
* At Airbnb, we have also have thousands of static libraries and one 3rd party dynamic framework all in one app.
* Using static libraries, we can strip all the exported symbols in the app binary. With the trick `-exported_symbols_list /dev/null`, we saved 14MB back then. Probably more nowadays.
* In the particular case that we need to share code between the main app and the extension, we found duplicating the static libs actually adds less size than using a dylib. It's because we have lots of small modules, which create lots of public symbols, which need to be preserved as exported symbols in the final dylib binary. Also, thanks to the super long Swift symbols, ridiculously, the symbols are bigger than the code. 
* We use the "mono dylib" approach in our Focus project for debugging. People specify which modules they need to work on before generating the workspace. Leveraging the dependency graph, we are able to figure out which modules they won't be able to change in this workspace. Then those modules are grouped in one dylib.

**Reply by tinder-maxwellelliott** [reply]
> We use the "mono dylib" approach in our Focus project for debugging. People specify which modules they need to work on before generating the workspace. Leveraging the dependency graph, we are able to figure out which modules they won't be able to change in this workspace. Then those modules are grouped in one dylib.

Our developers are very against the multiple focused project approach, preferring one "God" project (global search and edit anything are key drivers). How did you nudge people to use this explicit target open approach?

**Reply by keith** [reply]
How is the general performance of the giant project? That was the biggest thing that convinced devs to use the smaller ones. It has its downsides with choosing what should be included, and having to regenerate if you're wrong.

But also focused projects vs the mono dylib don't particularly have to be intertwined

**Reply by tinder-maxwellelliott** [reply]
Performance is very poor in the large project, full deadlock is not uncommon, but developers are not confused on where sources could live and what they can edit.

**Comment by qyang-nj** [comment]
> How did you nudge people to use this explicit target open approach?

We don't need to nudge people, because the alternative, opening the "God" project, is waaaay worse. Xcode is consistently  unresponsive in the "God" project.
We also tried to make it's easy for people to select modules, like provide an option to focus on my team modules. Our `focus` command also supports pipe operation. It can be combined with any other commands producing file paths, like `git`, `grep`, `find`, etc.

**Reply by qyang-nj** [reply]
(I guess I typed in the wrong place. I intended to reply this in the above thread 😅).

**Comment by mwaterfall** [comment]
It looks like static linking is prominently used here, and it's interesting to hear about projects with hundreds if not thousands of libs.

I'm curious to hear a little bit more about dependency graphs with this number of static libs.

In our experience, for example, when we have UI libraries used by multiple feature libraries and also within the main app target, we're seeing symbol duplications causing runtime issues when conditionally type casting views loaded from nibs, and just general duplicate symbol warnings.

When it comes to the myriad number of common capabilities/behaviours in our app, that feature libraries also need to adopt whilst on the journey of modularisation, I'm surprised to hear that dynamic libs are not used a lot heavier, and I'd be keen to hear some of the approaches used to work around this.

For example, when splitting up common things like logging, analytics, UI frameworks, localisation, general utilities, helpers and extensions, the main target and also many many other libs need access to these definitions, which when all linked together at built time has caused these symbol duplication and runtime type casting issues.

I can see that decoupling libs and reducing dependencies by instead building against common interfaces for things like logging and analytics will allow the removal of direct target dependencies. The "wiring up" can then be done in the main app, but for things like utility and helper types, classes and views and all the other common models that aren't easy to sit behind a layer of abstraction, it feels like these all need to be in a dynamic library that most libs and the main app will link against.

The problem there is that to have proper separation of concerns, it would make sense to keep code separated and to keep interfaces and helpers isolated into their own modules/libs, which would then start increasing the number of dynamic libs and hence increase launch times. Unless of course these modules all sit under one "core/common" dynamic lib, but then you have the issue of increasing build times when you need to recompile the full common dynamic lib if you just change a small piece of one interface.

I'd be grateful if anyone could share their insights into their dependency graphs and how that all fits into their static vs dynamic linking setup 🙏🏻

**Reply by shahzadmajeed** [reply]
You probably want to take a look at following 2 articles:

But the TL;DR of my separate experiments is that both all dynamic & all static are better approaches than hybrid setup. All static (using all targets as static including third party dependencies) is definitely the best option so far. The build is optimized for both speed (startup times) and size in this case. I control all this with Tuist.

1. https://www.emergetools.com/blog/posts/static-vs-dynamic-frameworks-ios-discussion-chat-gpt
2. https://bpoplauschi.github.io/2021/10/25/Advanced-static-vs-dynamic-libraries-and-frameworks.html

However, [mergable libraries](https://developer.apple.com/documentation/xcode/configuring-your-project-to-use-mergeable-libraries) in Xcode 15 is an interesting concept where Apple is promising best of both worlds i.e., builds optimized for speed and size. We will see how that unwraps.

**Reply by aehlke** [reply]
Any newer insights since the past couple Xcode releases?
